### ISSUE TITLE
Change `end_offset` type to `timedelta` in `TimeWindowPartitionsDefinition`

### ISSUE THREAD

**Issue by paulius-hydrosat** [dagster-io/dagster#15923]
### What's the use case?

My concrete usecase is to materialize daily partitions with a 5 day delay. This use case is supported in theory, but unusable in practice due to the slow path in `_get_last_partition_window` (https://github.com/dagster-io/dagster/blob/9660abf4a7b138eddb907816976252913a092cfa/python_modules/dagster/dagster/_core/definitions/time_window_partitions.py#L478.)

Instead of trying to fix the slow path, I would like to suggest making `end_offset` more flexible and clearer.

### Ideas of implementation

I suggest that end_offset should become a timedelta and shift both:

* time when time window partitions start existing
* time when time window partitions are scheduling

The docstring could read as follows:

```
end_offset (timedelta): Shifts the schedule by the indicated offset. E.g. if
    schedule_offset is timedelta(hours=1) then partitions will be materialized 1 hour past
    the end of the time window.
```

The implementation would consist of three major parts:

1. replacing `current_time` with `current_time - end_offset` in various methods of `TimeWindowPartitionsDefinition`.
2. implementing "shifted" scheduler
3. finding the best way to transition from `int` to `timedelta`.

The first part should be simple, and the resulting implementation would be a simplification from the current one.

The second part may be tricky, I have not looked at it yet. From a first glance, shifting a cron based schedule should be straightforward except for the directives like `*/15`.

For the third part there are several options how to approach it, but I don't know which one would be better.

### Additional information

If you think this is worthwhile, I could contribute the implementation.

### Message from the maintainers

Impacted by this issue? Give it a 👍! We factor engagement into prioritization.

**Comment by smackesey** [comment]
cc @sryza @clairelin135

**Comment by sryza** [comment]
I don't think we can change this, for backcompat reasons, but we could support both

**Comment by paulius-hydrosat** [comment]
I had the following idea: expand the type of `end_offset` to `int | timedelta` (currently it is `int`) and if `int` is passed, then convert it to corresponding `timedelta` object based on the window size. If window size is not constant, use any of the sizes and raise warning suggesting to use `timedelta`.

This should result in a rather small breakage, mainly in the case of non-constant window size and obscure corner cases listed in #15306 (not generating partitions when start_date is in the future and not removing partitions when end_date is specified).
